platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -1,73 +1,393 @@
+            vcom: false,
+        }))
+        .await;
+
+        let (cmd_prod, cmd_cons) = KChannel::new_async(1).await.split();
+        let commander = CommanderTask {
+            cmd: cmd_cons,
+            display_info: DisplayInfo {
+                frames,
+                frame_idx: 0,
+            },
+            ctxt: ctxt.clone(),
+        };
+
+        let vcom = VCom {
+            kernel,
+            ctxt: ctxt.clone(),
+        };
+        let draw = Draw {
+            kernel,
+            buf: linebuf,
+            spim: SpiSenderClient::from_registry(kernel).await.unwrap(),
+            ctxt,
+        };
+
+        kernel.spawn(commander.run()).await;
+        kernel.spawn(vcom.run()).await;
+        kernel.spawn(draw.run()).await;
+
+        kernel
+            .with_registry(|reg| reg.register_konly::<EmbDisplayService>(&cmd_prod))
+            .await
+            .map_err(|_| FrameError::DisplayAlreadyExists)?;
+
+        Ok(())
     }
+}
 
-    k.sleep(Duration::from_millis(100)).await;
+/// One entire frame, stored one bit per pixel
+struct FullFrame {
+    frame: [[u8; WIDTH_BYTES]; HEIGHT],
+}
 
-    // Loop, toggling the VCOM
-    let mut vcom = true;
-    let mut ctr = 0u32;
-    let mut linebuf = FixedVec::new((52 * 240) + 2).await;
-    for _ in 0..(52 * 240) + 2 {
-        let _ = linebuf.try_push(0);
+impl FullFrame {
+    pub fn new() -> Self {
+        Self {
+            frame: [[0u8; WIDTH_BYTES]; HEIGHT],
+        }
     }
+}
 
-    loop {
-        // It takes ~50ms to send a full frame, and 20fps is every
-        // 66.6ms. TODO: once we have "intervals", use that here.
-        k.sleep(Duration::from_millis(17)).await;
+impl FullFrame {
+    #[inline]
+    fn set_px(&mut self, x: usize, y: usize, color: Gray8) {
+        if x >= WIDTH || y > HEIGHT {
+            return;
+        }
+        let byte_x = x / 8;
+        let bit_x = x % 8;
 
-        // Send a pattern
-        //
-        // A note on this format:
-        //
-        // * Every FRAME gets a 1 byte command.
-        //     * It is zero, unless we are toggling VCOM.
-        //     * VCOM must be toggled once per second...ish.
-        // * Foreach LINE (240x) - 52 bytes total:
-        //     * 1 byte for line number
-        //     * (400bits / 8 = 50bytes) of data (one bit per pixel)
-        //     * 1 "dummy" byte
-        // * At the END, we need a total of two dummy bytes, so one from the last line, + 1 more
-        //
-        // This is where the 52x240 + 1 + 1 buffer size comes from.
+        if color.luma() > 128 {
+            self.frame[y][byte_x] |= 1 << (bit_x as u8);
+        } else {
+            self.frame[y][byte_x] &= !(1 << (bit_x as u8))
+        }
+    }
+}
+
+impl Dimensions for FullFrame {
+    fn bounding_box(&self) -> embedded_graphics::primitives::Rectangle {
+        Rectangle::new(
+            Point { x: 0, y: 0 },
+            Size {
+                width: 400,
+                height: 240,
```

Comment:
nit: should we use the constants here?
```suggestion
                width: WIDTH,
                height: HEIGHT,
```
(we might have to cast from `usize` to whatever these fields are, i think they might be `u32` but can't remember OTTOMH?)